### PR TITLE
Bug 2247735: Adding toleration to schedule CSI pods on master nodes

### PR DESCRIPTION
**Pull request by rchikatw**
Adding toleration to schedule CSI pods on master nodes

### CODE REVIEW

File: pkg/utils/k8sutils.go [hunk 1]
```diff
@@ -36,6 +36,8 @@ const StorageClientNamespaceEnvVar = "STORAGE_CLIENT_NAMESPACE"
 
 const StatusReporterImageEnvVar = "STATUS_REPORTER_IMAGE"
 
+const CSIDaemonSetsTolerateAllNodes = "CSI_DaemonSets_Tolerate_All_Nodes"
```
Comment (Madhu-1) [inline pkg/utils/k8sutils.go:39]:
```suggestion
const runCSIDaemonsetOnMaster = "RUN_CSI_DAEMONSET_ON_MASTER"
```

File: pkg/utils/placements.go [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+	"strconv"
+
+	corev1 "k8s.io/api/core/v1"
+)
+
+func GetTolerationForCSIPods() corev1.Toleration {
+	enableCSIPodSchedulingOnMaster, err := strconv.ParseBool(os.Getenv(CSIDaemonSetsTolerateAllNodes))
+	if err != nil {
+		log.Fatal(err)
+	}
```
Comment (Madhu-1) [inline pkg/utils/placements.go:15]:
```suggestion
// Run on master by default
runOnMaster :=true
var err error
rom:=os.Getenv(runCSIDaemonsetOnMaster))
if rom!=""{
runOnMaster, err = strconv.ParseBool(rom)
	if err != nil {
		log.Fatal(err)
	}
}

if runOnMaster{
	toleration := corev1.Toleration{
			Key:      "node-role.kubernetes.io/master",
			Operator: corev1.TolerationOpExists,
			Value:    "true",
			Effect:   corev1.TaintEffectNoSchedule,
		}
		return toleration
	]
}
```

File: pkg/utils/placements.go [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+	rom := os.Getenv(runCSIDaemonsetOnMaster)
+	if rom != "" {
+		runOnMaster, err = strconv.ParseBool(rom)
+		if err != nil {
+			log.Fatal(err)
+		}
+	}
+	if runOnMaster {
+		toleration := corev1.Toleration{
+			Key:      "node-role.kubernetes.io/master",
```
Comment (Madhu-1) [inline pkg/utils/placements.go:24]: it looks like this is deprecated and replaced with control-plane https://kubernetes.io/docs/reference/labels-annotations-taints/#node-role-kubernetes-io-master-taint, please check what we have in supported OCP versions

File: pkg/utils/placements.go [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+			Operator: corev1.TolerationOpEqual,
+			Value:    "true",
```
Comment (Madhu-1) [inline pkg/utils/placements.go:26]: value is not required as we will not have it set for the master node and also the Operator should be exists not equal. please check https://docs.openshift.com/container-platform/4.14/nodes/scheduling/nodes-scheduler-taints-tolerations.html

**Review by Madhu-1 — COMMENTED** [review]
please test and change and see if its working as expected

**Comment by leelavg** [comment]
@Madhu-1 and @rchikatw 
- I'm wondering why not `runCSIDaemonSetOnAllNodes` and use `Toleration{Operator: TolerationExists}`

Reasoning being,
- no need to find and set the `Toleration.Key` explicitly.
- Only if workloads are scheduled on tainted nodes then only CSI pods does actual work (this is a con as well, they are always there consuming resources)
- let the default be false (or not set) to conform to existing deployments and take the input from Subscription `.spec.config.env` (or what is the planned deployment model, iirc env will be reset if not set in CSV?)

**Comment by Madhu-1** [comment]
> @Madhu-1 and @rchikatw
> 
> * I'm wondering why not `runCSIDaemonSetOnAllNodes` and use `Toleration{Operator: TolerationExists}`
> 

The requirement was to run daemonset pods also on master node but not the nodes which might have tainted for something else.

> Reasoning being,
> 
> * no need to find and set the `Toleration.Key` explicitly.
> * Only if workloads are scheduled on tainted nodes then only CSI pods does actual work (this is a con as well, they are always there consuming resources)
> * let the default be false (or not set) to conform to existing deployments and take the input from Subscription `.spec.config.env` (or what is the planned deployment model, iirc env will be reset if not set in CSV?)

👍🏻  default is `true` but it need to be added to env to avoid upgrade problem

File: pkg/utils/placements.go (hunk 3 above)
Comment (rchikatw) [reply pkg/utils/placements.go:24]: checked one of the fusion hci clusters the key was "node-role.kubernetes.io/master". we can go with this key

**Comment by leelavg** [comment]
@rchikatw could you pls fix commitlint https://github.com/red-hat-storage/ocs-client-operator/actions/runs/6930384492/job/18849951078#step:4:25?

**Review by leelavg — APPROVED** [review]

**Pull request merged**